tests/garage/tf/models/test_model.py

```diff
@@ -0,0 +1,216 @@
+        input_var = tf.placeholder(tf.float32, shape=(None, 5))
+        model = SimpleModel(output_dim=2, name="MySimpleModel")
+        outputs = model.build(input_var, name='network_2')
+        data = np.ones((3, 5))
+        result, result2 = self.sess.run(
+            [outputs, model.networks['network_2'].outputs],
+            feed_dict={model.networks['network_2'].input: data})
+        assert np.array_equal(result, result2)
+        assert model.name == "MySimpleModel"
+
+    def test_same_model_with_no_name(self):
+        input_var = tf.placeholder(tf.float32, shape=(None, 5))
+        another_input_var = tf.placeholder(tf.float32, shape=(None, 5))
+        model = SimpleModel(output_dim=2)
+        model.build(input_var)
+        with self.assertRaises(ValueError):
+            model.build(another_input_var)
+
+        model2 = SimpleModel(output_dim=2)
+        with self.assertRaises(ValueError):
+            model2.build(another_input_var)
+
+    def test_model_with_different_name(self):
+        input_var = tf.placeholder(tf.float32, shape=(None, 5))
+        another_input_var = tf.placeholder(tf.float32, shape=(None, 5))
+        model = SimpleModel(output_dim=2)
+        outputs_1 = model.build(input_var)
+        outputs_2 = model.build(another_input_var, name='network_2')
+        data = np.ones((3, 5))
+        results_1, results_2 = self.sess.run([outputs_1, outputs_2],
+                                             feed_dict={
+                                                 input_var: data,
+                                                 another_input_var: data
+                                             })  # noqa: E126
+        assert np.array_equal(results_1, results_2)
+
+    def test_model_with_different_name_in_different_order(self):
+        input_var = tf.placeholder(tf.float32, shape=(None, 5))
+        another_input_var = tf.placeholder(tf.float32, shape=(None, 5))
+        model = SimpleModel(output_dim=2)
+        outputs_1 = model.build(input_var, name='network_1')
+        outputs_2 = model.build(another_input_var)
+        data = np.ones((3, 5))
+        results_1, results_2 = self.sess.run([outputs_1, outputs_2],
+                                             feed_dict={
+                                                 input_var: data,
+                                                 another_input_var: data
+                                             })  # noqa: E126
+        assert np.array_equal(results_1, results_2)
+
+    def test_model_in_model(self):
+        input_var = tf.placeholder(tf.float32, shape=(None, 5))
+        model = ComplicatedModel(output_dim=2)
+        outputs = model.build(input_var)
+        data = np.ones((3, 5))
+        out, model_out = self.sess.run(
+            [outputs, model.networks['default'].outputs],
+            feed_dict={model.networks['default'].input: data})
+
+        assert np.array_equal(out, model_out)
+
+    def test_model_pickle(self):
```

Comment:
did you address KR's concerns about the pickling procedure?

Reply:
This looks right to me. Anson and I discussed it at some length.